src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs

```diff
@@ -19,66 +19,647 @@
+
+            var result = finder.GetDirectories(baseDir, pattern);
+            var actual = result.Select(x => x.FullName);
+
+            CollectionAssert.AreEquivalent(expected, actual);
+            baseDir.Parent.DidNotReceive().GetDirectories(Arg.Any<string>(), Arg.Any<SIO.SearchOption>());
+            baseDir.Received().GetDirectories(pattern, SIO.SearchOption.TopDirectoryOnly);
+        }
+
+        [TestCase("tests/v*")]
+        [TestCase("tests/*")]
+        [TestCase("tests/v?")]
+        [TestCase("*/v*")]
+        [TestCase("*/v?")]
+        [TestCase("**/v*")]
+        [TestCase("**/v?")]
+        [TestCase("te*/v*")]
+        [TestCase("te*/*")]
+        [TestCase("te*/v?")]
+        [TestCase("t?sts/v*")]
+        [TestCase("t?sts/*")]
+        [TestCase("t?sts/v?")]
+        [TestCase("./tests/v*")]
+        [TestCase("./tests/*")]
+        [TestCase("./tests/v?")]
+        [TestCase("./*/v*")]
+        [TestCase("./*/v?")]
+        [TestCase("./**/v*")]
+        [TestCase("./**/v?")]
+        [TestCase("./te*/v*")]
+        [TestCase("./te*/*")]
+        [TestCase("./te*/v?")]
+        [TestCase("./t?sts/v*")]
+        [TestCase("./t?sts/*")]
+        [TestCase("./t?sts/v?")]
+        [TestCase("**/tests/v*")]
+        [TestCase("**/tests/*")]
+        [TestCase("**/tests/v?")]
+        [TestCase("**/*/v*")]
+        [TestCase("**/*/v?")]
+        [TestCase("**/te*/v*")]
+        [TestCase("**/te*/*")]
+        [TestCase("**/te*/v?")]
+        [TestCase("**/t?sts/v*")]
+        [TestCase("**/t?sts/*")]
+        [TestCase("**/t?sts/v?")]
+        public void GetDirectories_MultipleComponents_MultipleMatches(string pattern)
+        {
+            var finder = new DirectoryFinder(this.fileSystem);
+            var baseDir = this.GetFakeDirectory("tools", "metamorphosator");
+            var testsDir = this.GetFakeDirectory("tools", "metamorphosator", "tests");
+            baseDir.GetDirectories("tests", SIO.SearchOption.TopDirectoryOnly).Returns(new IDirectory[] { testsDir });
+            baseDir.GetDirectories("te*", SIO.SearchOption.TopDirectoryOnly).Returns(new IDirectory[] { testsDir });
+            baseDir.GetDirectories("t?sts", SIO.SearchOption.TopDirectoryOnly).Returns(new IDirectory[] { testsDir });
+            testsDir.GetDirectories("v*", SIO.SearchOption.TopDirectoryOnly).Returns(new IDirectory[] { this.GetFakeDirectory("tools", "metamorphosator", "tests", "v1"), this.GetFakeDirectory("tools", "metamorphosator", "tests", "v2") });
+            testsDir.GetDirectories("*", SIO.SearchOption.TopDirectoryOnly).Returns(new IDirectory[] { this.GetFakeDirectory("tools", "metamorphosator", "tests", "v1"), this.GetFakeDirectory("tools", "metamorphosator", "tests", "v2") });
+            testsDir.GetDirectories("v?", SIO.SearchOption.TopDirectoryOnly).Returns(new IDirectory[] { this.GetFakeDirectory("tools", "metamorphosator", "tests", "v1"), this.GetFakeDirectory("tools", "metamorphosator", "tests", "v2") });
```

Comment:
Perhaps this would be easier to read if we extracted `new IDirectory[] { this.GetFakeDirectory("tools", "metamorphosator", "tests", "v1"), this.GetFakeDirectory("tools", "metamorphosator", "tests", "v2") }` into a variable?

Reply:
Yes. I extracted the content of `testsDir` as suggested to a variable. Did the same for `baseDir` too...